truffle/test/controller_deployment_interaction.js

```diff
@@ -0,0 +1,330 @@
+
+    } catch(e) {
+      
+      return assert(e.message.includes('_secondsToExpiration'));
+
+    }
+
+    return assert(false);
+
+  });
+
+  it("Fail to create loan with less than a unit of DAI", async function () {
+
+    try {
+
+      loanObject._requestedValueAttoDai = BigInt(Math.pow(10,17));
+
+      await tuiChainController.createLoan(...Object.values(loanObject));
+
+    } catch(e) {
+      
+      return assert(e.message.includes('VM Exception while processing transaction: revert'));
+
+    }
+
+    return assert(false);
+
+  });
+
+  it("Fail to create loan without a multiple integer of a unit of DAI", async function () {
+
+    try {
+
+      loanObject._requestedValueAttoDai = 1.5*Math.pow(10,18);
+
+      await tuiChainController.createLoan(...Object.values(loanObject));
+
+    } catch(e) {
+      
+      return assert(e.message.includes('overflow'));
+
+    }
+
+    return assert(false);
+
+  });
+
+  it("Fail to create a loan if not the owner", async function () {
+
+    try {
+    
+      await tuiChainController.createLoan(...Object.values(loanObject), {from: accounts[1]});
+
+    } catch(e) {
+
+      return assert(e.message.includes('caller is not the owner'));
+
+    }
+
+  });
+  
+  it("Create a loan with a requested DAI value of 1000 DAI", async function () {
+
+    const transaction = await tuiChainController.createLoan(...Object.values(loanObject));
+
+    tuiChainLoan = await TuiChainLoan.at(transaction.logs[0].address);
+
+    return assert((await tuiChainLoan.getToken) != 0x0000000000000000000000000000000000000000);
+
+  });
+
+  /* -------------------------------------------------------------------------- */
+
+  it("Fail to finalize the loan if not the owner", async function () {
+
+    try {
+
+      assert((await tuiChainLoan.checkExpiration.call()) == false);
+    
+      await tuiChainController.finalizeLoan(tuiChainLoan.address, {from: accounts[1]});
+
+    } catch(e) {
+
+      return assert(e.message.includes('caller is not the owner'));
+
+    }
+
+  });
+
+  it("Fail to finalize loan before it turns Active", async function () {
+
+    try {
+
+      assert((await tuiChainLoan.checkExpiration.call()) == false);
+
+      await tuiChainController.finalizeLoan(tuiChainLoan.address);
+
+      return assert(false);
+
+    } catch(e) {
+
+      return assert(true);
+
+    }
+
+  });
+
+  it("Finalize the loan", async function () {
+
+    // 10 accounts giving 100 DAI each
+    for (let index = 0; index < accounts.length; index++) {
+      
+      await daiMock.increaseAllowance(tuiChainLoan.address, BigInt(200* Math.pow(10,18)), {from: accounts[index]});
+      await tuiChainLoan.provideFunds(BigInt(100 * Math.pow(10,18)), {from: accounts[index]});  
+      
+    }
+
+    // check that loan is ACTIVE 
+    var events = await tuiChainLoan.getPastEvents('PhaseUpdated', {fromBlock: 0, toBlock: 'latest'});
+
+
+    assert(events[0].returnValues.newPhase == 3); // 3 is the ACTIVE phase in the loan contract enum
```

Comment:
Doesn't something like `TuiChainLoan.Phase` exist?

Reply:
Enums are not supported by the ABI. It's complicated but for now we have to stick with this tightly coupled version of test things.

Reply:
Makes sense. I would create an enum in the test file nonetheless (no enums in JS but cf. https://stackoverflow.com/a/5040502/4551082).

Reply:
ok I can do that